File: Code/encryption.py
"""
    Performs the encryption algorithm for the key exchange
"""


def encrypt(data, key, progressFunction=None,
            progressData=None, progressInterval=250000):
    newData = []

    origKey = key
    key = key % 2560

    key = key*2

    rOrig = (key*10)**key % 123
    r = rOrig
    yMax = len(data) - 1
    y = 0
    x = 0

    while y <= yMax:
        c = data[y]
        newData.append((chr((ord(c) + r % 256) % 256)))
        r *= (key + 1+int(r/key)) % 250
        if r >= 10000000 or r <= 0:
            r = rOrig

        if (x >= progressInterval):
            x = 0
            if progressFunction is not None:
                progressFunction(y, yMax, origKey, progressData)

        y = y + 1
        x = x + 1

    if progressFunction is not None:
        progressFunction(y, yMax, origKey, progressData)

    return newData


def decrypt(data, key, progressFunction=None,
            progressData=None, progressInterval=1000000):
    newData = []

    origKey = key
    key %= 2560

    key *= 2

    rOrig = (key*10)**key % 123
    r = rOrig
    yMax = len(data) - 1
    y = 0
    x = 0

    while y <= yMax:
        oldVal = ord(data[y])

        oldVal -= r % 256

        if oldVal < 0:
            oldVal += 256

        newData += (chr(oldVal))

        r *= (key + 1+int(r/key)) % 250

        if r >= 10000000 or r <= 0:
            r = rOrig

        if (x >= progressInterval):
            x = 0
            if progressFunction is not None:
                progressFunction(y, yMax, origKey, progressData)

        y = y + 1
        x = x + 1

    return newData

File: OLD/NodeJS Standard Client/encryption.js
var bigInt = require("big-integer");
var util = require("./utilities.js")

function CarterEncrypt(data, key) {
    var newData = ""
    key = key % 2560;
    key = key*2;
    key = bigInt(key);
    var r = key.multiply(10);
    r = r.pow(key).mod(123);
    var rOrig = r;
  
    var y = 0;
    var x = 0;
    var yMax = data.length - 1;
  
    while (y <= yMax) {
      var characterInt = util.utf16ToDig(data.charAt(y));
      newData += util.intToChar(bigInt(characterInt).add(r.mod(256)).mod(256));
      
      var factor = key.add(1);
      var factor2 = r.divide(key);
      factor = factor.add(factor2);
      factor = factor.mod(250);
  
      r = r.multiply(factor)
  
      if (r >= 10000000 || r <= 0) {
        r = rOrig;
      }
  
      y = y + 1;
      x = x + 1;
    }
  
    return newData;
  }

  
  function CarterDecrypt(data, key) {
    var newData = ""
    key = key % 2560;
    key = key*2;
    key = bigInt(key);
    var r = key.multiply(10);
    r = r.pow(key).mod(123);
    var rOrig = r;
  
    var y = 0;
    var x = 0;
    var yMax = data.length - 1;
  
    while (y <= yMax) {
      var oldVal = bigInt(util.utf16ToDig(data.charAt(y)));
      oldVal = oldVal.minus(r.mod(bigInt(256)));
      oldVal = Number(oldVal.toString());
  
      if (oldVal < 0) {
        oldVal = oldVal+256;
      }
  
      newData = newData + util.intToChar(oldVal);
    
      var factor = key.add(1);
      var factor2 = r.divide(key);
      factor = factor.add(factor2);
      factor = factor.mod(250);
  
      r = r.multiply(factor)
      if (r >= 10000000 || r <= 0) {
        r = rOrig
      }
      y = y + 1;
      x = x + 1;
    }
  
    return newData;
  }

  module.exports = {
      encrypt: CarterEncrypt,
      decrypt: CarterDecrypt
  }